src/System.Resources.Extensions/ref/System.Resources.Extensions.cs

```diff
@@ -25,13 +25,14 @@ public sealed partial class PreserializedResourceWriter : System.Resources.IReso
         public void AddResource(string name, System.IO.Stream value, bool closeAfterWrite = false) { }
         public void AddResource(string name, object value) { }
         public void AddResource(string name, string value) { }
+        public void AddResource(string name, string typeName, string value) { }
```

Comment:
@tarekgh @rainersigwald  I'd like your feedback on these two API changes before submitting for approval.

I could have completed this work without API changes.  

I could have kept the `AddTypeConverterResource` method named the same and have it handle primitives, but it felt incorrect since it would do deserialization and add primitive resources.  To me this is the "core" API of the type, so giving it a better name felt more correct.

I could have allowed a null `typeName` in `AddBinaryFormattedResource`, but this would be the only API that defines behavior for null in the entire assembly, so I felt an overload was better.

Reply:
Are we adding this API to handle the primitive types only? or you think we may use it in the future to handle more non primitive types? I am asking because if we limit this API for the primitive type, we may choose more strict method name to set the expectation of the caller.

Also, did we think to have this API takes type converter instead of the type name? I am asking because this may add more flexibility handling more types through the converter and not inside the API itself.

---
In reply to: [288560944](https://github.com/dotnet/corefx/pull/38012#discussion_r288560944) [](ancestors = 288560944)

Reply:
> I could have allowed a null typeName in AddBinaryFormattedResource, but this would be the only API that defines behavior for null in the entire assembly, so I felt an overload was better.


Actually I prefer to have 
AddBinaryFormattedResource(string name, string typeName, byte[] value) 

takes default type value to null 

AddBinaryFormattedResource(string name, byte[] value, string typeName = null) 

so the caller don't have to pass null explicitly. I think it is ok to have one of the APIs handle null value instead of exposing a new API but I am not strong on that opinion. 
The drawback from my proposal of the default type parameter is the parameters types/names order will not be consistent with the other APIs.


---
In reply to: [289064287](https://github.com/dotnet/corefx/pull/38012#discussion_r289064287) [](ancestors = 289064287,288560944)

Reply:
> Are we adding this API to handle the primitive types only

It's actually meant to handle any type.  For primitive types we will convert them, for non-primitive types we persist the string value.

> did we think to have this API takes type converter instead of the type name?0

If it took a type converter that would imply that the target Type was activated in the same process as the Writer.  In the general case that is not what we want.  We only special case activating primitive types because we know they are safe and can represent them more concisely in binary form.

> so the caller don't have to pass null explicitly.

The current overload does not require the caller to pass null explicitly.

> parameters types/names order will not be consistent with the other APIs.

Exactly.  I felt that the natural ordering for APIs that take a type would be for the type to precede the value, since that matches what you'd be writing in a declaration.  



---
In reply to: [289066428](https://github.com/dotnet/corefx/pull/38012#discussion_r289066428) [](ancestors = 289066428,289064287,288560944)